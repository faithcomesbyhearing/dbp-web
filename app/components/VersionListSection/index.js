import React from 'react';
import PropTypes from 'prop-types';
import {
	Accordion,
	AccordionItem,
	AccordionItemPanel,
	AccordionItemHeading,
	AccordionItemButton,
} from 'react-accessible-accordion';

class VersionListSection extends React.PureComponent {
	render() {
		const { items } = this.props;

		return (
			<Accordion>
				{items.map((item) => {
					const { key, title, className, text, altText, types, clickHandler } =
						item;

					if (types.audio && types.audio_drama) {
						return (
							<AccordionItem key={key} className="accordion-title-style">
								<AccordionItemHeading>
									<AccordionItemButton>
										<h4 title={title} className={className}>
											{altText ? `${text} (${altText})` : text}
										</h4>
									</AccordionItemButton>
								</AccordionItemHeading>
								<AccordionItemPanel className="accordion-body-style">
									<a
										role="button"
										key={`${key}_drama`}
										tabIndex={0}
										className="version-item-button"
										onClick={() => clickHandler('audio_drama')}
										type="button"
									>
										Dramatized Version
									</a>
									<a
										role="button"
										key={`${key}_plain`}
										tabIndex={0}
										className="version-item-button"
										onClick={() => clickHandler('audio')}
										type="button"
									>
										Non-Dramatized Version
									</a>
								</AccordionItemPanel>
							</AccordionItem>
						);
					}

					return (
						<AccordionItem key={key} className="accordion-title-style">
							<AccordionItemHeading aria-level={1}>
								<AccordionItemButton>
									<a
										role="button"
										key={key}
										tabIndex={0}
										title={title}
										className={`${className} top-level-title`}
										onClick={() => clickHandler('')}
										type="button"
									>
										{altText ? `${text} (${altText})` : text}
									</a>
								</AccordionItemButton>
							</AccordionItemHeading>
							<AccordionItemPanel />
						</AccordionItem>
					);
				})}
			</Accordion>
		);
	}
}

VersionListSection.propTypes = {
	items: PropTypes.arrayOf(
		PropTypes.shape({
			key: PropTypes.string.isRequired,
			title: PropTypes.string.isRequired,
			className: PropTypes.string,
			text: PropTypes.string.isRequired,
			altText: PropTypes.string,
			types: PropTypes.shape({
				audio: PropTypes.bool,
				audio_drama: PropTypes.bool,
				text_plain: PropTypes.bool,
				text_format: PropTypes.bool,
			}).isRequired,
			clickHandler: PropTypes.func.isRequired,
		}),
	),
};

export default VersionListSection;
