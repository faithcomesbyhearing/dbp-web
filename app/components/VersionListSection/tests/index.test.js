import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { fromJS } from 'immutable';
import VersionListSection from '..'; // Assuming the path is correct
import {
	itemsParser,
	filterFunction,
	getTexts,
} from './versionListSectionUtils';

const filterText = '';
const activeTextId = 'ENGESV';
const activeBookId = 'MAT';
const activeChapter = 1;

const items = [
	{
		path: { textId: 'ENGESV', bookId: 'MAT', chapter: 1 },
		key: 'ENGESV2001',
		className: '',
		title: 'English Standard Version',
		text: 'English Standard Version',
		altText: '',
		types: {
			audio_drama: true,
			text_plain: true,
			audio: true,
			text_format: true,
		},
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGNIV', bookId: 'MAT', chapter: 1 },
		key: 'ENGNIV1978',
		className: 'active-version',
		title: 'New International Version',
		text: 'New International Version',
		altText: '',
		types: {
			text_plain: true,
			audio: true,
			audio_drama: true,
			video_stream: true,
		},
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGKJV', bookId: 'MAT', chapter: 1 },
		key: 'ENGKJV1611',
		className: '',
		title: 'King James Version',
		text: 'King James Version',
		altText: '',
		types: { audio: true, audio_drama: true, text_plain: true },
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGNAB', bookId: 'MAT', chapter: 1 },
		key: 'ENGNAB1970',
		className: '',
		title: 'New American Bible',
		text: 'New American Bible',
		altText: '',
		types: { text_format: true, audio_drama: true, text_plain: true },
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGWEB', bookId: 'MAT', chapter: 1 },
		key: 'ENGWEB1997',
		className: '',
		title: 'World English Bible',
		text: 'World English Bible (Hosanna audio)',
		altText: 'World English Bible',
		types: { text_plain: true, text_format: true, audio_drama: true },
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGWWH', bookId: 'MAT', chapter: 1 },
		key: 'ENGWWH1997',
		className: '',
		title: 'World English Bible (Afred Henson)',
		text: 'World English Bible (Afred Henson)',
		altText: '',
		types: { audio_drama: true, text_plain: true },
		clickHandler: jest.fn(),
	},
	{
		path: { textId: 'ENGNIVA', bookId: 'MAT', chapter: 1 },
		key: 'ENGNIVAnull',
		className: '',
		title: 'New International Version (Anglicised)',
		text: 'New International Version (Anglicised)',
		altText: '',
		types: { audio_drama: true, audio: true, text_plain: true },
		clickHandler: jest.fn(),
	},
];

describe('<VersionListSection />', () => {
	it('Should match previous snapshot with valid props', () => {
		const { asFragment, queryAllByText } = render(<VersionListSection items={items} />);

		// Check if accordion items are rendered
		expect(queryAllByText('Non-Dramatized Version').length).toEqual(4); // Update the number based on the expected count of buttons
		expect(queryAllByText('Dramatized Version').length).toEqual(4); // Update the number based on the expected count of buttons

		// Ensure the snapshot matches
		expect(asFragment()).toMatchSnapshot();
	});

	it('Should match previous snapshot using data from API', async () => {
		const apiItems = await getTexts({ languageCode: 6414 });

		const sectionItems = itemsParser(
			fromJS(apiItems),
			activeTextId,
			activeBookId,
			activeChapter,
			filterText,
			filterFunction,
			(bible, audioType) => `${bible}_${audioType}`,
		);

		const { asFragment } = render(<VersionListSection items={sectionItems} />);

		// Ensure the snapshot matches
		expect(asFragment()).toMatchSnapshot();
	});

	it('Should handle clicking on audio drama button', () => {
		const { queryAllByText } = render(<VersionListSection items={items} />);

		// Simulate clicking the "Dramatized Version" button
		const audioDramaButtons = queryAllByText('Dramatized Version');
		fireEvent.click(audioDramaButtons[0]);

		// Verify that the clickHandler was called
		expect(items[0].clickHandler).toHaveBeenCalledWith('audio_drama');
	});

	it('Should handle clicking on non-dramatized version button', () => {
		const { queryAllByText } = render(<VersionListSection items={items} />);

		// Simulate clicking the "Non-Dramatized Version" button
		const nonDramatizedButtons = queryAllByText('Non-Dramatized Version');
		fireEvent.click(nonDramatizedButtons[0]);

		// Verify that the clickHandler was called
		expect(items[0].clickHandler).toHaveBeenCalledWith('audio');
	});

	it('Should match previous snapshot using data from api and language ID 17045 English: USA', async () => {
		const apiItems = await getTexts({ languageCode: 17045 });

		const sectionItems = itemsParser(
			fromJS(apiItems),
			activeTextId,
			activeBookId,
			activeChapter,
			filterText,
			filterFunction,
			(bible, audioType) => `${bible}_${audioType}`,
		);

		const { container } = render(<VersionListSection items={sectionItems} />);
		const bibles = container.querySelectorAll('.accordion-body-style');
		expect(bibles.length).toEqual(3);
	});
});
