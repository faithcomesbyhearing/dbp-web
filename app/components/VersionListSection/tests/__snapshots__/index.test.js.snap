// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionListSection /> Should match previous snapshot using data from api 1`] = `
<div
  className="accordion"
  role="tablist"
>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-7"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-7"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className="active-version"
        title="English Standard Version"
      >
        English Standard Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-7"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-7"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-8"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-8"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="New International Version"
      >
        New International Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-8"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-8"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-9"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-9"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="King James Version"
      >
        King James Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-9"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-9"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-10"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-10"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="New American Bible"
      >
        New American Bible
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-10"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-10"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-11"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-11"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="World English Bible (Hosanna audio)"
      >
        World English Bible (Hosanna audio)
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-11"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-11"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-12"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-12"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="World English Bible (Afred Henson)"
      >
        World English Bible (Afred Henson)
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-12"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-12"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-13"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-13"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="New International Version (Anglicised)"
      >
        New International Version (Anglicised)
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-13"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-13"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
</div>
`;

exports[`<VersionListSection /> Should match previous snapshot with valid props 1`] = `
<div
  className="accordion"
  role="tablist"
>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-0"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-0"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="English Standard Version"
      >
        English Standard Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-0"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-0"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-1"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-1"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className="active-version"
        title="New International Version"
      >
        New International Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-1"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-1"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-2"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-2"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="King James Version"
      >
        King James Version
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-2"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-2"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-3"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-3"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="New American Bible"
      >
        New American Bible
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-3"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-3"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-4"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-4"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="World English Bible"
      >
        World English Bible (Hosanna audio) ( World English Bible )
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-4"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-4"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-5"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-5"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <a
        className=" top-level-title"
        onClick={[Function]}
        role="button"
        tabIndex={0}
        title="World English Bible (Afred Henson)"
      >
        World English Bible (Afred Henson)
      </a>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-5"
      className="accordion__body accordion__body--hidden"
      id="accordion__body-5"
      role="tabpanel"
    />
  </div>
  <div
    className="accordion-title-style"
  >
    <div
      aria-controls="accordion__body-6"
      aria-selected={false}
      className="accordion__title"
      disabled={false}
      id="accordion__title-6"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="tab"
      tabIndex="0"
    >
      <h4
        className=""
        title="New International Version (Anglicised)"
      >
        New International Version (Anglicised)
      </h4>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="accordion__title-6"
      className="accordion-body-style accordion__body--hidden"
      id="accordion__body-6"
      role="tabpanel"
    >
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Dramatized Version
      </a>
      <a
        className="version-item-button"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Non-Dramatized Version
      </a>
    </div>
  </div>
</div>
`;
